driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java

```diff
@@ -292,7 +292,7 @@ private void getMore() {
                 throw translateCommandException(e, serverCursor);
             }
             resourceManager.setServerCursor(nextServerCursor);
-            if (limitReached()) {
+            if (limitReached() || !resourceManager.operable()) {
```

Comment:
I don't think there is currently a bug as specified in JAVA-5214, nor that this check is needed. Following are my observations, though if I am missing something, please correct me.

Currently, `ResourceManager.execute` guarantees that both server and client resources are released (via `ResourceManager.doClose`) if `QueryBatchCursor.close` is called concurrently with `QueryBatchCursor.hasNext`/`next`/`tryNext`. And if `QueryBatchCursor.close` is ordered either before or after (in the synchronization order, which in this particular case also implies the happens-before order) `QueryBatchCursor.hasNext`/`next`/`tryNext`, then `QueryBatchCursor.close` itself releases those resources (again, via `ResourceManager.doClose`).

More specifically, here the `!resourceManager.operable()` check was added to the `getMore` method. `getMore` is executed by `ResourceManager.execute`, which calls `endOperation` in a `finally` block, which in turn checks the state and calls `doClose` if needed. See also VAKOTODO on why the modified `should close cursor after getMore finishes if cursor was closed while getMore was in progress and getMore returns a response` fails.

P.S. You may see that `doHasNext` contains the `!resourceManager.operable()` check, but that is only because it calls `getMore` in a loop, and the loop must not continue if `QueryBatchCursor.close` as called. Note how `doHasNext` does nor releases anything itself, because it by design relies on `ResourceManager.execute` doing that.

P.P.S. The only time when `QueryBatchCursor` calls `ResourceManager.releaseServerAndClientResources` itself, instead of relying on `ResourceManager.execute`, is when `limitReached()`, which is again by design: `ResourceManager.execute` does not understand the semantics of the "operation" it executes, and can't know that despite `QueryBatchCursor.close` not having been called, the cursor may release resources because no more data is going to be fetched. This is just an optimization, and it had already been there before the introduction of `ResourceManager`.